Partial content

damonkelley opened the pull request:
These changes add support for range requests.

I added a function `binary-slurp-range` that can handle the different position offsets of a Range header value. (i.e. `-500`, `9500-`, and `0-500`). The `file` handler is now checking if the request is a range request, and serving the partial content, or serving the whole file.

I also updated the request parser to parse the range field-value into a map

``` clojure
{:units "bytes" :start 0 :end 50}
{:units "bytes" :start nil :end 50}
```
### Cob Spec

🎉 Partial Content! 🎉 

![image](https://cloud.githubusercontent.com/assets/4121849/18280683/8257393a-741f-11e6-946e-92617d7733d6.png)

damonkelley committed:
Move the file filesystem handlers to a new namespace

damonkelley committed:
Cleanup file spec

damonkelley committed:
Add binary-slurp-range to the file namespace

damonkelley committed:
Parse the Range header when the request is created

damonkelley committed:
Modify binary-slurp-range to handle nil offsets

damonkelley committed:
Update range header to allow empty position offsets

damonkelley committed:
Serve partial content of files if the range header is present

trptcolin commented on spec/http_clj/file_spec.clj:
```diff
@@ -3,23 +3,44 @@
+               (seq (file/binary-slurp @test-path)))))
+
+  (context "binary-slurp-range"
+    (before (spit @test-path "Read a range of bytes"))
```
It's interesting to think about the implications of non-ASCII data, specifically data that takes multiple bytes to encode. Not a bug or missing test or anything, but the fact that you're grabbing part of a string made me think about what'd happen if the first character was multi-byte. And then how that'd change our perception of the correct thing. 

Like for example, [the range request spec](https://tools.ietf.org/html/rfc7233#section-2.2) specifies that we ought to be OCP about range types, and that makes me think we should be able to ask for characters if we want. Not a requirement for this story, but that might be kind of an interesting cob_spec addition down the road if you don't already have one in mind.

So this is totally fine for this story/PR - you've definitely implemented byte ranges - just thinking about making it more open to extension in the future.

trptcolin commented on spec/http_clj/request/parser_spec.clj:
```diff
@@ -34,4 +34,10 @@
+               (parser/parse-field-values {:content-length "10" :host "www.example.com"})))
+
+    (it "parses Range"
+      (should= {:range {:units "bytes" :start 0 :end 4}}
```
Nice that you've represented the `:units` in the data structure, so I suspect you've already given some thought to what I was rambling on about regarding OCP.

trptcolin commented on src/http_clj/file.clj:
```diff
@@ -9,6 +10,30 @@
+    (with-open [stream (io/input-stream path)]
+      (.skip stream start)
+      (.read stream buffer))
+    buffer))
```
You sure like pattern-matching 😉  

These look good. You can also get a bit more pattern-matching fanciness w/ [core.match](https://github.com/clojure/core.match) if you feel like you need something more than just class-based matching.

trptcolin commented on src/http_clj/request/parser.clj:
```diff
@@ -26,11 +26,38 @@
+(defn- parse-start-end [range-map]
+  (-> range-map
+      (update :start parse-byte-position)
+      (update :end parse-byte-position)))
```
I've seen you use `update` a bunch - really interesting - it wasn't in the language until fairly recently so I haven't found myself using it. I'll have to keep my eyes open for opportunities now.

trptcolin commented on src/http_clj/file.clj:
```diff
@@ -9,6 +10,30 @@
+  (let [buffer-size (inc (- end start))
+        buffer (byte-array buffer-size)]
+    (with-open [stream (io/input-stream path)]
+      (.skip stream start)
```
Nice use of `.skip`. 

I'd be interested in knowing what happens if someone passes a range that's outside of what the legitimate indexes for the resource in question. Like if I read bytes 999999991-999999999 of a 10-byte file, what happens - does it fail?

damonkelley replied:
Yeah, in the scenario you describe, it returns `[0 0 0 0 0 0 0 0 0]`. If a range like 0-500 was passed in, it returns all 10 bytes of the file, and pads the rest of the buffer with `0`s.

damonkelley replied:
It seems like an exception should be thrown if such a range is provided. Within the request handler we can then catch the exception and return a `416 Range Not Satisfiable`

trptcolin replied:
👍  agreed on the right solution there.

trptcolin commented:
👍  apart from the one question on what happens if the given range is... out of range?

damonkelley committed:
Raise an exception if the range is not satisfiable

damonkelley committed:
Extract a partial-file handler

This handler is used when a range header is present. If an exception is
raised when reading the specified range, a 416 response is sent back to
the client.

damonkelley committed:
Refactor the file handler to accept a path

damonkelley commented on src/http_clj/file.clj:
```diff
@@ -9,6 +10,39 @@
+  [path start end]
+  (when (not (valid-range? path start end))
+    (throw (ex-info "Range Unsatisfiable" {})))
+  (-binary-slurp-range path start end))
```
@trptcolin Now throwing an exception if the range isn't valid
